### ISSUE TITLE
Add AppLabelNameSet and VersionLabelNameSet Inputs

### ISSUE THREAD

**Issue by wyattfry** [kiali/kiali#7603]
### What do you want to improve?

I want to improve the flexibility of the app and version label names.

### What is the current behavior?

Kiali supports only one label key for app and version.

### What is the new behavior?

I would like to have the option of specifying multiple app and version label names. Istio supports both `app`/`version` and the longer `app.kubernetes.io/name`/`app.kubernetes.io/version` but Kiali does not.

Documentation addressing this limitation: https://kiali.io/docs/configuration/istio/#labels-and-resource-names

> Although Istio lets you use different labels on different pods, Kiali can only use a single set.

Or to take it a step further, add the ability to make a "best guess" as to the proper values if the labels are missing. E.g.
- App could default to `Pod.OwnerReferences[0].Name` or if all else fails, `Pod.Name`
- Version could default to `split(Pod.Containers[*].Image, ":")[1]` where the container is the one specified by the annotation `kubectl.kubernetes.io/default-container` or if no such annotation exists, just picking the first one

**Comment by jmazzitelli** [comment]
> I would like to have the option of specifying multiple app and version label names

We thought about doing this 5 years ago, but it was decided against as the performance hit would be too much (it involves having to perform multiple k8s queries for everything). Read [this issue for context](https://github.com/istio/istio/issues/10724) and [my comment in its abandoned PR here](https://github.com/istio/istio/pull/10807#issuecomment-458659223) (as well as the follow-up comments from others in that PR). My comment I linked to there explains the issue why supporting multiple labels is no good.

Being able to auto-discover what they should be is an interesting thought. Kiali will still only support one set, but it might be interesting to discuss how it might be possible to try to guess what they should be if they are not defined in the Kiali config.

**Comment by wyattfry** [comment]
Thanks for the explanation; I figured there was a specific reason why multiple label support was not implemented. Maybe the the auto discover could be gated behind an option? What would the process be for seeing if the discover feature is viable?

**Comment by jmazzitelli** [comment]
Just off the top of my head, I would think the auto-discovery could be done in the server. The server helm chart and operator probably will only care about setting them in the kiali configmap for the server to consume - I do not think the helm charts nor operator actually use the app/version labels for anything (I'm not 100% sure of that, but I think that is true). If it is true, then the helm charts and operator don't have to worry about doing any auto-discovery (unless, for some reason, the auto-discovery method requires permissions that the server doesn't have).

Assuming that's all good, suppose then the server starts up and sees there is no app or version label defined. That should trigger the server to perform the auto-discovery. The question then becomes, WHAT is the server going to examine to try to discover the labels? Look at the Istio control plane namespace for something in there? Does Istio have a configuration in its own ConfigMap (which the kiali server has access to and can read) that defines what it uses for app and version?

That's the big question - WHAT is analyzed to determine the app / version labels to use. What's the algorithm or formula, for lack of a better word.

@nrfox this goes along with the work you were doing wrt auto-discovering things so we don't have to explicitly configure them in the Kiali CR. This could potentiall be another set of values in the Kiali CR that we can auto-discover and not force the user to set. I think this is going to be harder to auto-discover, but I think its a good conversation to have, in case it actually is possible to do it effectively.

**Comment by wyattfry** [comment]
Yeah, that is THE question, my first attempt was something like this:

- App could default to `Pod.OwnerReferences[0].Name` or if all else fails, `Pod.Name`
- Version could default to `split(Pod.Containers[*].Image, ":")[1]` where the container is the one specified by the annotation `kubectl.kubernetes.io/default-container` or if no such annotation exists, just picking the first one

**Comment by nrfox** [comment]
> @nrfox this goes along with the work you were doing wrt auto-discovering things so we don't have to explicitly configure them in the Kiali CR. This could potentiall be another set of values in the Kiali CR that we can auto-discover and not force the user to set. I think this is going to be harder to auto-discover, but I think its a good conversation to have, in case it actually is possible to do it effectively.

What is being autodiscovered? The app label for a single app or for every application in the mesh.

> but it was decided against as the performance hit would be too much (it involves having to perform multiple k8s queries for everything)

Maybe this was a concern back then but it's no longer much of a concern since all kube objects Kiali needs to query are already cached so querying them is very cheap.

> It's a helm chart of charts, a couple dozen "primary" subcharts with a bunch more extended add on charts. We are having to go through each subchart and surgically insert the kiali-requreid labels, and each chart is a little different, so it's kinda tedious.

That doesn't sound fun.

Since queries are cheap, can we just query them in the same order of precedence that istio does? https://istio.io/latest/docs/ops/deployment/application-requirements/#pod-requirements. I'm not totally sure the implications of this inside Kiali but that would be my preference. Then there's also no need for the Kiali `app_label_name` and `version_label_name` config options.

**Comment by jmazzitelli** [comment]
> Since queries are cheap, can we just query them in the same order of precedence that istio does? https://istio.io/latest/docs/ops/deployment/application-requirements/#pod-requirements. I'm not totally sure the implications of this inside Kiali but that would be my preference. Then there's also no need for the Kiali `app_label_name` and `version_label_name` config options.

Right. That's what we looked at 5 years ago. It would be difficult for me to vote for Kiali to change so that it does 3 queries EVERY time we want a resource, regardless of it coming from the cache or not. At minimum, doing performance tests to ensure there really won't be a performance hit would need to be done.

Maybe we do that as a fallback behavior if the Kiali CR does not explicitly set `app_label_name`/`version_label_name`:

1. Is there "service.istio.io/canonical-name" for the resource in question? Then use the "service.istio.io" labels
2. Else - is there "app.kubernetes.io/name" for the resource in question? Then use "app.kubernetes.io" labels
3. Else - use "app"/"version" labels

We'd have to make some kind of opinionated rule to say something like "you can't use a combination of different labels" - for example, if you use service.istio.io for name, you need it for version, too. Otherwise, we're now talking about doing a potential max of 6 queries lol. In other words, we shouldn't bother to support a resource that combines two forms; for example, a resource with `service.istio.io/canonical-name` for its name and `app.kubernetes.io/version` (or `version`) for its version (and you just know somebody out there is doing that :)). They would need to configure the currently supported Kiali CR settings `app_label_name` and `version_label_name` for that.

That could be something to look at. It would help people who have different apps with different labels (albeit with potential performance issues), but people who have a consistent labeling scheme would be able to tell Kiali a more efficient way to do things (how Kiali works today).

I also don't know the implication of this for the graph generator (which has to deal with Prometheus queries, in addition to k8s queries). That would need to be investigated, too.

**Comment by nrfox** [comment]
> We'd have to make some kind of opinionated rule to say something like "you can't use a combination of different labels"

I don't think we need to do this. We just query them in the same order of precedence that Istio does and we only accept the labels that Istio has in their pod requirements. It shouldn't matter if they mix and match as long as one of them matches the labels defined in Istio pod requirements.

> I also don't know the implication of this for the graph generator (which has to deal with Prometheus queries, in addition to k8s queries). That would need to be investigated, too.

Yes that's the part I am unsure of as well. iirc graph queries use `canonical-name` and `canonical-revision` exclusively but don't quote me on that.

**Comment by wyattfry** [comment]
> Maybe we do that as a fallback behavior if the Kiali CR does not explicitly set app_label_name/version_label_name:
>
> 1. Is there "service.istio.io/canonical-name" for the resource in question? Then use the "service.istio.io" labels
> 1. Else - is there "app.kubernetes.io/name" for the resource in question? Then use "app.kubernetes.io" labels
> 1. Else - use "app"/"version" labels

Even this change would be a huge help. Many helm charts we use already have `app.kubernetes.io/name|version` labels, a smaller number have the (older?) `app|version`, and a few have neither.

I think not supporting mixed styles within a pod is very reasonable.

**Comment by jshaughn** [comment]
This has been a pain point for a long time.  I don't love that there are probably many touch points to impl an optioned approach, but maybe it's time to take the hit, given that perf is no longer a major concern.  If users specify label names then we use them.  If not we apply the same approach as Istio, which will handle mixed-name scenarios.

I don't think it's a concern for the graph, because metrics already use "canonical" fields set by istio.  The final question I have is whether this affects trace querying, because if it does that may be a non-starter, we can't afford multiple trace queries.

**Comment by jshaughn** [comment]
Moving to sprint to first understand what sort of impact this would have on the code.  If we can centralize things we may be able to move forward.  Prior perf concerns may have been mitigated by our recent caching improvements.

**Comment by jshaughn** [comment]
After a cursory look at the code, I don't think this carries over to trace fetches, so that's good news. I will also follow up with @josunect .

But, I'm going to dive into this a bit more, code-wise.

**Comment by josunect** [comment]
> After a cursory look at the code, I don't think this carries over to trace fetches, so that's good news. I will also follow up with [@josunect](https://github.com/josunect) .
> 
> But, I'm going to dive into this a bit more, code-wise.

For tracing, the `app` label is used: see https://github.com/kiali/kiali/blob/master/business/workloads.go#L2526

**Comment by jshaughn** [comment]
> For tracing, the app label is used: see https://github.com/kiali/kiali/blob/master/business/workloads.go#L2526

Hmm, that's maybe a little unfortunate but maybe not.  If we are just trying to fetch traces for a single workload, and the config leaves the appLabelName unset, we could possibly just examine the labels on that workload, and use the one we find (it should be one of 3 possibilities).

**Comment by wyattfry** [comment]
Thanks!

**Comment by jshaughn** [comment]
@wyattfry You're welcome, this has been a high-want for a long time.  The change will be in v2.6. The changes hit a lot of places, so please let us know if you see anything strange.  Note that for existing users that want to run a mixed scheme, the config map would need to be updated to unset `istio_labels.app_label_name` and `istio_labels.version_label_name` (right, @jmazzitelli ?).
